### PR TITLE
Add different integration approximation methods

### PR DESCRIPTION
**Pull request by iwei20**
I wrote some methods of more accurate numerical integration, given data points, in SLMath.  I also modified PIDController to use this new integrate(), but only with the current and previously measured errors (which is why I used trapezoidal).  I think that maybe if we store some extra error points we could use Simpson.

### CODE REVIEW

**Comment by selym3** [comment]
Might want to make the integrator filter a variable of PIDController, in the same way `IFilter mIFilter` is and has the method `PIDController setIntegratorFilter(IFilter filter)`. A new setter could be used to make the integral Riemann by default in the constructor.

Also try moving the code that gets you the resutl of `SLMath.integrate(...)` (including the arrays and prevError stuff) into a private helper function. This might make `calcuate(...)` easier to understand.

Also, I don't think your hasPrevError works the way im guessing it should. In the constructor it's set to true, and the ternary operator might be flipped on line 77.

**Review by selym3 — COMMENTED** [review]
Might want to make the integrator filter a variable of PIDController, in the same way `IFilter mIFilter` is and has the method `PIDController setIntegratorFilter(IFilter filter)`. A new setter could be used to make the integral Riemann by default in the constructor.

Also try moving the code that gets you the resutl of `SLMath.integrate(...)` (including the arrays and prevError stuff) into a private helper function. This might make `calcuate(...)` easier to understand.

Also, I don't think your hasPrevError works the way im guessing it should. In the constructor it's set to true, and the ternary operator might be flipped on line 77.

**Comment by iwei20** [comment]
By the way, these changes are to increase the accuracy of integral calculation.

File: src/com/stuypulse/stuylib/control/PIDController.java [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 import com.stuypulse.stuylib.streams.filters.IFilter;
 import com.stuypulse.stuylib.math.SLMath;
+import com.stuypulse.stuylib.math.SLMath.Integral;
 
 /**
  * This PID controller is built by extending the Controller class. It has a dynamic rate, so it can
```
Comment (Sam-Belliveau) [inline src/com/stuypulse/stuylib/control/PIDController.java:8]: Ivan pls add author tag

File: src/com/stuypulse/stuylib/math/SLMath.java [hunk 2]
```diff
@@ -145,6 +145,128 @@ public static double circular(double x) {
...
+     * 
+     * Arrays x and y must be the same length, because they contain data points.
+     * The x-values must be in sorted order.
+     * 
+     * @param x       the x-values of the data points, where (x[i], y[i]) is a point
+     * @param y       the y-values of the data points, where (x[i], y[i]) is a point
+     * @param formula the chosen formula to approximate the integral.
+     * @return approximation of the integral from given points
+     */
+    public static double integrate(double[] x, double[] y, Integral formula) {
```
Comment (Sam-Belliveau) [inline src/com/stuypulse/stuylib/math/SLMath.java:163]: This is the type of input that can be done with a vector2d

File: src/com/stuypulse/stuylib/math/SLMath.java [hunk 3]
```diff
@@ -145,6 +145,128 @@ public static double circular(double x) {
...
+        }
+
+        // Precondition: x-values should be sorted
+        for(int i = 1; i < x.length; ++i) {
+            if(x[i] < x[i - 1]) {
+                throw new IllegalArgumentException("x-values should be in sorted order");
+            }
+        }
+
+        return formula.calculate(x, y);
```
Comment (Sam-Belliveau) [inline src/com/stuypulse/stuylib/math/SLMath.java:176]: I understand that the x values need to be sorted  but I think theres a better way

File: src/com/stuypulse/stuylib/math/SLMath.java [hunk 4]
```diff
@@ -145,6 +145,128 @@ public static double circular(double x) {
...
+
+    /**
+     * Integral is an interface used by integrate() to get the method used for numerical integration.  
+     * It also contains several classes containing various formulas for integration.
+     * 
+     * Riemann: Uses the Riemann sum to approximate the integral. (https://en.wikipedia.org/wiki/Riemann_sum)<br>
+     * Trapezoidal: Uses the trapezoidal rule to approximate the integral. (https://en.wikipedia.org/wiki/Trapezoidal_rule)<br>
+     * Simpson: Uses Composite Simpson's Rule for irregularly spaced data to approximate the integral. (https://en.wikipedia.org/wiki/Simpson%27s_rule)<br>
+     */
+    public interface Integral {
```
Comment (Sam-Belliveau) [inline src/com/stuypulse/stuylib/math/SLMath.java:199]: Great idea for a class, you should make it its own file,

File: src/com/stuypulse/stuylib/math/SLMath.java [hunk 5]
```diff
@@ -145,6 +145,128 @@ public static double circular(double x) {
...
+        /**
+         * Given data points, gives an approximation for the integral.
+         * Arrays x and y must be the same length, because they contain data points.
+         * The x-values must be in sorted order.
+         * 
+         * @param x the x-values of the data points, where (x[i], y[i]) is a point
+         * @param y the y-values of the data points, where (x[i], y[i]) is a point
+         * @return approximation of the integral from given points
+         */
+        public double calculate(double[] x, double[] y);
```
Comment (Sam-Belliveau) [inline src/com/stuypulse/stuylib/math/SLMath.java:209]: Vector2D instead of X, Y. Also it might be better if the integral keeps an internal list, then you can call a .update(Vector2D) to give it a new value, I feel like it would make more sense for an integral class. It would remove the need for the array

however, you could still use an array through the use of a static helper function that could call update on each item of the array

This would makes its use in PID easier because you wouldnt need an integral double, boolean for checking if it has the past error, and it would just be a lot easier.

this would also make using the simpson method of integrals possible in PID

File: src/com/stuypulse/stuylib/math/SLMath.java [hunk 6]
```diff
@@ -145,6 +145,128 @@ public static double circular(double x) {
...
+                double trapezoidsum = 0.0;
+                for(int i = 1; i < x.length; ++i) {
+                    trapezoidsum += (y[i] + y[i - 1]) * (x[i] - x[i - 1]) / 2;
+                }
+                return trapezoidsum;
+            }
+        }
+
+        public class Simpson implements Integral {
+            public double calculate(double[] x, double[] y) {
```
Comment (Sam-Belliveau) [inline src/com/stuypulse/stuylib/math/SLMath.java:232]: I really lke your implementations, they are clean, pls put your author tag on these,

**Review by Sam-Belliveau — COMMENTED** [review]
tldr; a couple changes to the structure of the integral classes

maybe put the classes in a folder in the math folder? I think maybe the integral class can store its own internal list? I think that would be good, however its up to you where you want to go from here. I think this is significant enough to be seperate from SLMath.

File: src/com/stuypulse/stuylib/math/SLMath.java [hunk 7]
```diff
@@ -145,6 +145,128 @@ public static double circular(double x) {
...
+                // Create an array of interval widths
+                double[] widths = new double[x.length - 1];
+                for(int i = 0; i < x.length - 1; ++i) {
+                    widths[i] = x[i + 1] - x[i];
+                }
+
+                // Even case for intervals
+                int n = widths.length;
+                for(int i = 0; i < n / 2; i += 1) {
+                    double ai = (2 * Math.pow(widths[2 * i + 1], 3) - Math.pow(widths[2 * i], 3) + 3 * widths[2 * i] * Math.pow(widths[2 * i + 1], 2)) /
```
Comment (Sam-Belliveau) [inline src/com/stuypulse/stuylib/math/SLMath.java:244]: Use SLMath.fpow() here because its a lot faster when using integer powers

Comment (iwei20) [reply src/com/stuypulse/stuylib/math/SLMath.java:244]: Oops I forgot regular pow wasn't binary exponentiation

File: src/com/stuypulse/stuylib/math/SLMath.java (hunk 6 above)
Comment (iwei20) [reply src/com/stuypulse/stuylib/math/SLMath.java:232]: myles helped a bit with recommending how i should structure
